File: app/src/main/java/opencrowd/hgc/hgcwallet/hapi/tasks/UpdateBalanceTaskAPI.java
package opencrowd.hgc.hgcwallet.hapi.tasks;

import com.hederahashgraph.api.proto.java.Query;
import com.hederahashgraph.api.proto.java.Response;
import com.hederahashgraph.api.proto.java.ResponseCodeEnum;

import java.util.Date;
import java.util.List;

import opencrowd.hgc.hgcwallet.hapi.APIRequestBuilder;
import opencrowd.hgc.hgcwallet.hapi.APIBaseTask;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.database.DBHelper;
import opencrowd.hgc.hgcwallet.database.account.Account;

public class UpdateBalanceTaskAPI extends APIBaseTask {

    public UpdateBalanceTaskAPI() {
        super();
    }

    @Override
    public void main() {
        super.main();
        List<Account> accounts = DBHelper.getAllAccounts();
        if (accounts != null && !accounts.isEmpty()) {
            Account payer = accounts.get(0);
            if (payer.accountID() == null) {
                return;
            }
            for (Account account : accounts) {
                if (account.accountID() != null) {
                    Query query = APIRequestBuilder.requestForGetBalance(payer, account.accountID(), node.accountID());
                    try {
                        log(query.toString());
                        Response response = cryptoStub().cryptoGetBalance(query);
                        log(response.toString());
                        ResponseCodeEnum precheck = response.getCryptogetAccountBalance().getHeader().getNodeTransactionPrecheckCode();
                        switch (precheck) {
                            case OK:
                                long balance = response.getCryptogetAccountBalance().getBalance();
                                account.setBalance(balance);
                                account.setLastBalanceCheck(new Date());
                                DBHelper.saveAccount(account);
                                break;
                            case INVALID_ACCOUNT_ID:
                                Singleton.INSTANCE.clearAccountData(account);
                                error = "Invalid account " + account.accountID().stringRepresentation();
                                break;

                            default:
                                error = Singleton.INSTANCE.getErrorMessage(precheck);
                                break;
                        }


                    } catch (Exception e) {
                        e.printStackTrace();
                        error = getMessage(e);
                    }
                }
            }
        }
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/hapi/tasks/UpdateBalanceTaskAPI.kt
package hedera.hgc.hgcwallet.hapi.tasks

import com.hederahashgraph.api.proto.java.ResponseCodeEnum
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.database.DBHelper
import hedera.hgc.hgcwallet.database.account.Account
import hedera.hgc.hgcwallet.hapi.APIBaseTask
import hedera.hgc.hgcwallet.hapi.GetBalanceParam
import hedera.hgc.hgcwallet.hapi.TransactionBuilder
import hedera.hgc.hgcwallet.modals.HGCAccountID
import java.util.*


class UpdateBalanceTaskAPI(val accounts: List<Account>) : APIBaseTask() {

    override fun main() {
        super.main()
        if (accounts.isNotEmpty()) {

            val errorList = mutableListOf<String>()
            val payerAccount = HGCAccountID.fromString("0.0.0")!!

            accounts.forEach { account ->
                account.accountID()?.let { accID ->
                    var err: String? = null
                    try {
                        val txnBuilder = TransactionBuilder(null, payerAccount)
                        val getBalanceParam = GetBalanceParam(accID)
                        val (query, response) = grpc.perform(getBalanceParam, txnBuilder)
                        val precheck = response.cryptogetAccountBalance.header.nodeTransactionPrecheckCode
                        when (precheck) {
                            ResponseCodeEnum.OK -> {
                                val balance = response.cryptogetAccountBalance.balance
                                account.balance = balance
                                account.lastBalanceCheck = Date()
                                DBHelper.saveAccount(account)
                            }
                            ResponseCodeEnum.INVALID_ACCOUNT_ID -> {
                                Singleton.clearAccountData(account)
                                err = "${Singleton.getErrorMessage(precheck)} ${accID.stringRepresentation()}"
                            }

                            else -> err = "${Singleton.getErrorMessage(precheck)} ${accID.stringRepresentation()}"
                        }


                    } catch (e: Exception) {
                        e.printStackTrace()
                        err = getMessage(e)
                    }
                    err?.let { errorList.add(err) }
                }
            }

            if (errorList.isNotEmpty())
                error = errorList.joinToString("\n")
        }
    }
}